# Comparación de análisis TMT: Interpolado vs Raw

Este notebook carga dos análisis TMT con diferentes configuraciones de interpolación.

In [1]:
import sys
sys.path.insert(0, '/home/gianluca/Research/datapruebas_analysis')

from pathlib import Path

import pandas as pd

from src import config
from src.visualization.compare_interpolation_analysis import (
    load_interpolated_tmt_trials_by_date,
    load_raw_tmt_trials_by_date,
    compare_metrics,
    INTERPOLATED_DATE,
    RAW_DATE,
)

In [2]:
# Cargar los dos DataFrames SIN filtrar por trials comunes
# Valida que cada análisis tenga la configuración esperada
df_interp = load_interpolated_tmt_trials_by_date(INTERPOLATED_DATE)
df_raw = load_raw_tmt_trials_by_date(RAW_DATE)

print(f"Interpolado ({INTERPOLATED_DATE}):")
print(f"  - Trials: {len(df_interp):,}")
print(f"  - Sujetos: {df_interp['subject_id'].nunique()}")

print(f"\nRaw ({RAW_DATE}):")
print(f"  - Trials: {len(df_raw):,}")
print(f"  - Sujetos: {df_raw['subject_id'].nunique()}")

Excluding 18 subjects without both PART_A and PART_B trials: ['030720fd-dc61-422b-bdf2-2db9457d2f1c', '1d0fc2b7-c2a2-43bd-b00c-ab6b0a7ff722', '1e8869b5-564e-4104-b7e7-84cb6a8edf11', '5f6e5399-a575-4a52-91a7-136356728a8a', '82f8adea-8cec-410e-a7b8-dc3396162402']...


Excluding 18 subjects without both PART_A and PART_B trials: ['030720fd-dc61-422b-bdf2-2db9457d2f1c', '1d0fc2b7-c2a2-43bd-b00c-ab6b0a7ff722', '1e8869b5-564e-4104-b7e7-84cb6a8edf11', '5f6e5399-a575-4a52-91a7-136356728a8a', '82f8adea-8cec-410e-a7b8-dc3396162402']...


Interpolado (2026-01-16_19-20-02):
  - Trials: 5,621
  - Sujetos: 366

Raw (2026-01-16_18-57-15):
  - Trials: 5,607
  - Sujetos: 366


In [3]:
df_interp.head()

,subject_id,trial_id,trial_type,is_valid,trial_order_of_appearance,speed_threshold,dispositivo,mano,dispositivo-config,alcohol-drogas,...,nationality,experiment_origin,device,hand,device_config,alcohol_drugs,treatment,pad_usage,final_comment,start_date
0,df0b8572-3570-4c43-ac9b-017fec2adf22,NEUROPRUEBAS_2,PART_A,True,2,0.198439,Mouse,Derecha,si,no,...,Argentino,neuropruebas,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,df0b8572-3570-4c43-ac9b-017fec2adf22,NEUROPRUEBAS_3,PART_B,True,3,0.198439,Mouse,Derecha,si,no,...,Argentino,neuropruebas,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,df0b8572-3570-4c43-ac9b-017fec2adf22,NEUROPRUEBAS_6,PART_A,True,6,0.198439,Mouse,Derecha,si,no,...,Argentino,neuropruebas,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
6,df0b8572-3570-4c43-ac9b-017fec2adf22,NEUROPRUEBAS_8,PART_A,True,8,0.198439,Mouse,Derecha,si,no,...,Argentino,neuropruebas,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
8,df0b8572-3570-4c43-ac9b-017fec2adf22,NEUROPRUEBAS_10,PART_A,True,10,0.198439,Mouse,Derecha,si,no,...,Argentino,neuropruebas,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [4]:
df_raw.head()

,subject_id,trial_id,trial_type,is_valid,trial_order_of_appearance,speed_threshold,dispositivo,mano,dispositivo-config,alcohol-drogas,...,nationality,experiment_origin,device,hand,device_config,alcohol_drugs,treatment,pad_usage,final_comment,start_date
0,df0b8572-3570-4c43-ac9b-017fec2adf22,NEUROPRUEBAS_2,PART_A,True,2,0.280745,Mouse,Derecha,si,no,...,Argentino,neuropruebas,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,df0b8572-3570-4c43-ac9b-017fec2adf22,NEUROPRUEBAS_3,PART_B,True,3,0.280745,Mouse,Derecha,si,no,...,Argentino,neuropruebas,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,df0b8572-3570-4c43-ac9b-017fec2adf22,NEUROPRUEBAS_6,PART_A,True,6,0.280745,Mouse,Derecha,si,no,...,Argentino,neuropruebas,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
6,df0b8572-3570-4c43-ac9b-017fec2adf22,NEUROPRUEBAS_8,PART_A,True,8,0.280745,Mouse,Derecha,si,no,...,Argentino,neuropruebas,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
8,df0b8572-3570-4c43-ac9b-017fec2adf22,NEUROPRUEBAS_10,PART_A,True,10,0.280745,Mouse,Derecha,si,no,...,Argentino,neuropruebas,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


## Análisis de trials que no están en ambos datasets

In [5]:
# Identificar trials en cada dataset
#total trials
total_trials = set(zip(df_interp['subject_id'], df_interp['trial_id'])) | set(zip(df_raw['subject_id'], df_raw['trial_id']))
trials_interp = set(zip(df_interp['subject_id'], df_interp['trial_id']))
trials_raw = set(zip(df_raw['subject_id'], df_raw['trial_id']))

# Trials comunes
common_trials = trials_interp & trials_raw

# Trials solo en interpolado
only_interp = trials_interp - trials_raw

# Trials solo en raw
only_raw = trials_raw - trials_interp

print(f"\nTotal trials: {len(total_trials):,}")
print(f"Trials comunes: {len(common_trials):,}")
print(f"Solo en interpolado: {len(only_interp):,}")
print(f"Solo en raw: {len(only_raw):,}")


Total trials: 5,631
Trials comunes: 5,597
Solo en interpolado: 24
Solo en raw: 10


In [6]:
# Filtrar trials que solo están en un dataset
df_only_interp = df_interp[
    df_interp.apply(lambda r: (r['subject_id'], r['trial_id']) in only_interp, axis=1)
].copy()

df_only_raw = df_raw[
    df_raw.apply(lambda r: (r['subject_id'], r['trial_id']) in only_raw, axis=1)
].copy()

print(f"Trials solo en interpolado: {len(df_only_interp)}")
print(f"Trials solo en raw: {len(df_only_raw)}")

Trials solo en interpolado: 24
Trials solo en raw: 10


In [7]:
df_only_interp

,subject_id,trial_id,trial_type,is_valid,trial_order_of_appearance,speed_threshold,dispositivo,mano,dispositivo-config,alcohol-drogas,...,nationality,experiment_origin,device,hand,device_config,alcohol_drugs,treatment,pad_usage,final_comment,start_date
72,1e200df4-7a6b-4aab-b96b-8e8c8497716a,NEUROPRUEBAS_14,PART_A,True,14,0.302310,Mouse,Derecha,si,no,...,Argentina,neuropruebas,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
138,9b8b167b-7e64-4347-a579-ad8c2ba4025f,NEUROPRUEBAS_20,PART_A,True,20,0.268376,Mouse,Derecha,si,no,...,Argentina,neuropruebas,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
280,b1368f7d-47b5-45dd-83f4-5167eca84158,NEUROPRUEBAS_2,PART_A,True,2,0.111954,padNotebook,Derecha,no,no,...,Argentino,neuropruebas,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
876,1aadc486-9800-46c6-93f6-13aa11168a00,NEUROPRUEBAS_12,PART_A,True,12,0.301455,Mouse,Derecha,si,no,...,Ecuatoriana,neuropruebas,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
929,b4dc8afd-c6ae-4335-9694-217a9e6f2e05,NEUROPRUEBAS_5,PART_B,True,5,0.395467,Mouse,Derecha,si,si,...,Mexico,neuropruebas,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
936,b4dc8afd-c6ae-4335-9694-217a9e6f2e05,NEUROPRUEBAS_12,PART_A,True,12,0.395467,Mouse,Derecha,si,si,...,Mexico,neuropruebas,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
967,32d611d1-ad54-4835-99ed-460a443df14c,NEUROPRUEBAS_10,PART_A,True,10,0.203738,Mouse,Derecha,si,no,...,Argentina,neuropruebas,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1349,26815196-070f-4f94-a765-5bbecabea52c,NEUROPRUEBAS_10,PART_A,True,10,0.334165,Mouse,Derecha,si,no,...,franco argentino,neuropruebas,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1443,6838a4c2-ce67-4bd1-931f-2227776ce2a4,NEUROPRUEBAS_14,PART_A,True,14,0.145882,Mouse,Derecha,si,no,...,Argentina,neuropruebas,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1821,b65df3e7-91c6-40b7-a732-49622bfa8684,NEUROPRUEBAS_10,PART_A,True,10,0.198239,padNotebook,Derecha,si,no,...,Colombiano,neuropruebas,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [8]:
df_only_raw

,subject_id,trial_id,trial_type,is_valid,trial_order_of_appearance,speed_threshold,dispositivo,mano,dispositivo-config,alcohol-drogas,...,nationality,experiment_origin,device,hand,device_config,alcohol_drugs,treatment,pad_usage,final_comment,start_date
377,leandrogori2000@gmail.com-tmt-plugin_112.csv,NEUROPRUEBAS_2,PART_A,True,2,0.309733,NaN,NaN,NaN,NaN,...,Argentina,neuropruebas,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
960,32d611d1-ad54-4835-99ed-460a443df14c,NEUROPRUEBAS_3,PART_B,True,3,0.323079,Mouse,Derecha,si,no,...,Argentina,neuropruebas,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3407,48482d6a-a3ca-4c42-bdab-9462f26e6ff8,NEUROPRUEBAS_17,PART_B,True,17,0.338657,Mouse,Derecha,si,no,...,Perú,neuropruebas,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3712,4657de88-e8fe-4eca-b171-366b1822d2b3,NEUROPRUEBAS_20,PART_A,True,20,0.480592,Mouse,Izquierda,no,no,...,Argentina,neuropruebas,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4826,b17c4fc1-95ed-47b9-8dd5-72ee8705a62c,NEUROPRUEBAS_21,PART_B,True,21,0.297352,Mouse,Derecha,si,no,...,Colombiana,neuropruebas,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
5046,3cf3fc83-1ee9-4389-9978-fa5845035019,NEUROPRUEBAS_21,PART_B,True,21,0.557009,Mouse,Derecha,si,no,...,peruana,neuropruebas,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
6304,2697d032-ba2c-45e4-8115-588455f5b491,NEUROPRUEBAS_10,PART_A,True,10,0.328362,Mouse,Derecha,si,no,...,Mexicana,neuropruebas,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
6612,64bbfefd-5852-4921-8c79-c4f708b9b60a,DATAPRUEBAS_4,PART_A,True,4,0.362631,NaN,NaN,NaN,NaN,...,Argentina,datapruebas,Mouse,Derecha,NaN,NaN,no,NaN,NaN,2024-07-13T16:24:35.058000+00:00
7196,961c65a9-5caa-43c3-8cc4-c4c5484cd1cc,DATAPRUEBAS_8,PART_A,True,8,0.379155,NaN,NaN,NaN,NaN,...,Argentina,datapruebas,padNotebook,Izquierda,NaN,NaN,no,NaN,NaN,2024-02-08T15:36:29.660000+00:00
7459,139efc97-c335-401c-a98f-dd3825ab2d87,DATAPRUEBAS_11,PART_B,True,11,0.277513,NaN,NaN,NaN,NaN,...,Argentina,datapruebas,Mouse,Derecha,NaN,NaN,no,NaN,NaN,2024-02-01T20:24:30.641000+00:00


## Investigación: ¿Por qué la interpolación agrega casos?

### Hipótesis
La interpolación modifica el cálculo de `speed_threshold` y la validación de trials. Vamos a verificar:
1. Si los sujetos de trials únicos tienen otros trials en ambos datasets
2. Si las diferencias están en el cálculo de speed_threshold o en la validación individual de trials

In [9]:
# Análisis de sujetos con trials únicos en interpolado
subjects_only_interp = df_only_interp['subject_id'].unique()
print(f"Sujetos con trials únicos en interpolado: {len(subjects_only_interp)}")

# Verificar cuántos trials tiene cada sujeto en ambos datasets
for subj in subjects_only_interp[:5]:  # Mostrar primeros 5
    interp_count = len(df_interp[df_interp['subject_id'] == subj])
    raw_count = len(df_raw[df_raw['subject_id'] == subj])
    unique_count = len(df_only_interp[df_only_interp['subject_id'] == subj])
    print(f"\n{subj[:30]}...")
    print(f"  Trials en interpolado: {interp_count}")
    print(f"  Trials en raw: {raw_count}")
    print(f"  Trials únicos en interp: {unique_count}")

Sujetos con trials únicos en interpolado: 21

1e200df4-7a6b-4aab-b96b-8e8c84...
  Trials en interpolado: 18
  Trials en raw: 17
  Trials únicos en interp: 1

9b8b167b-7e64-4347-a579-ad8c2b...
  Trials en interpolado: 18
  Trials en raw: 17
  Trials únicos en interp: 1

b1368f7d-47b5-45dd-83f4-5167ec...
  Trials en interpolado: 11
  Trials en raw: 10
  Trials únicos en interp: 1

1aadc486-9800-46c6-93f6-13aa11...
  Trials en interpolado: 12
  Trials en raw: 11
  Trials únicos en interp: 1

b4dc8afd-c6ae-4335-9694-217a9e...
  Trials en interpolado: 11
  Trials en raw: 9
  Trials únicos en interp: 2


In [10]:
# Análisis de sujetos con trials únicos en raw
subjects_only_raw = df_only_raw['subject_id'].unique()
print(f"Sujetos con trials únicos en raw: {len(subjects_only_raw)}")

for subj in subjects_only_raw[:5]:
    interp_count = len(df_interp[df_interp['subject_id'] == subj])
    raw_count = len(df_raw[df_raw['subject_id'] == subj])
    unique_count = len(df_only_raw[df_only_raw['subject_id'] == subj])
    print(f"\n{subj[:30]}...")
    print(f"  Trials en interpolado: {interp_count}")
    print(f"  Trials en raw: {raw_count}")
    print(f"  Trials únicos en raw: {unique_count}")

Sujetos con trials únicos en raw: 10

leandrogori2000@gmail.com-tmt-...
  Trials en interpolado: 19
  Trials en raw: 20
  Trials únicos en raw: 1

32d611d1-ad54-4835-99ed-460a44...
  Trials en interpolado: 7
  Trials en raw: 7
  Trials únicos en raw: 1

48482d6a-a3ca-4c42-bdab-9462f2...
  Trials en interpolado: 11
  Trials en raw: 12
  Trials únicos en raw: 1

4657de88-e8fe-4eca-b171-366b18...
  Trials en interpolado: 18
  Trials en raw: 18
  Trials únicos en raw: 1

b17c4fc1-95ed-47b9-8dd5-72ee87...
  Trials en interpolado: 18
  Trials en raw: 19
  Trials únicos en raw: 1


In [11]:
# Comparación de speed_thresholds entre datasets para sujetos comunes
common_subjects = set(df_interp['subject_id'].unique()) & set(df_raw['subject_id'].unique())
print(f"Sujetos comunes: {len(common_subjects)}")

# Obtener speed_threshold por sujeto (es constante por sujeto)
st_interp = df_interp.groupby('subject_id')['speed_threshold'].first()
st_raw = df_raw.groupby('subject_id')['speed_threshold'].first()

# Comparar
st_comparison = pd.DataFrame({
    'speed_threshold_interp': st_interp,
    'speed_threshold_raw': st_raw
}).dropna()

st_comparison['diff'] = st_comparison['speed_threshold_interp'] - st_comparison['speed_threshold_raw']
st_comparison['pct_diff'] = (st_comparison['diff'] / st_comparison['speed_threshold_raw']) * 100

print(f"\nEstadísticas de diferencia en speed_threshold:")
print(st_comparison['pct_diff'].describe())

Sujetos comunes: 366

Estadísticas de diferencia en speed_threshold:
count    366.000000
mean     -23.958405
std       12.555745
min      -70.361472
25%      -32.124851
50%      -22.801010
75%      -13.939129
max        0.261436
Name: pct_diff, dtype: float64


### Análisis de timestamps en datos crudos

Para entender por qué ciertos trials son rechazados al interpolar (trials únicos en raw), necesitamos examinar sus timestamps crudos.

In [12]:
import ast

def load_raw_trial_timestamps(subject_id: str, trial_id: str, experiment_origin: str) -> list:
    """Cargar timestamps crudos de un trial específico."""
    
    if experiment_origin == 'neuropruebas':
        # Para neuropruebas, el subject_id puede ser el nombre del archivo CSV
        base_path = Path('/home/gianluca/Research/datapruebas_analysis/data/raw/tmt/neuropruebas/subjects')
        
        # Buscar el archivo que corresponde al subject_id
        csv_files = list(base_path.glob('*.csv'))
        matching_file = None
        
        for f in csv_files:
            # El subject_id puede ser el nombre del archivo o un UUID dentro del archivo
            if subject_id in f.name or subject_id.endswith('.csv'):
                # Si el subject_id es el nombre del archivo
                if f.name == subject_id or f.name == f"{subject_id}.csv":
                    matching_file = f
                    break
        
        # Si no encontramos por nombre de archivo, buscar por UUID en los archivos
        if matching_file is None:
            for f in csv_files:
                try:
                    df = pd.read_csv(f, nrows=1)
                    # Algunos archivos usan 'subject_id' o similar
                    if 'subject_id' in df.columns and df['subject_id'].iloc[0] == subject_id:
                        matching_file = f
                        break
                except:
                    continue
        
        if matching_file is None:
            # Intentar con el UUID directamente
            potential_file = base_path / f"{subject_id}.csv"
            if potential_file.exists():
                matching_file = potential_file
        
        if matching_file is None:
            return None, f"No se encontró archivo para subject_id: {subject_id}"
        
        df = pd.read_csv(matching_file)
        
        # Filtrar por trial_type TMT y encontrar el trial específico
        tmt_rows = df[df['trial_type'].isin(['tmt-trial', 'tmt'])]
        
        # El trial_id es NEUROPRUEBAS_N donde N es trial_index + 1
        trial_index = int(trial_id.split('_')[1]) - 1  # Convertir a índice 0-based
        
        if trial_index >= len(tmt_rows):
            return None, f"Trial index {trial_index} fuera de rango (total: {len(tmt_rows)})"
        
        row = tmt_rows.iloc[trial_index]
        
        # Extraer cursor_time
        cursor_time_str = row.get('cursor_time')
        if pd.isna(cursor_time_str):
            return None, "cursor_time es NaN"
        
        try:
            cursor_time = ast.literal_eval(cursor_time_str)
            return cursor_time, None
        except:
            return None, f"Error parseando cursor_time: {cursor_time_str[:100]}..."
    
    return None, f"experiment_origin no soportado: {experiment_origin}"


# Analizar un trial que solo está en raw (debería tener timestamps duplicados)
sample_raw_only = df_only_raw.iloc[0]
print(f"Analizando trial solo en raw:")
print(f"  Subject: {sample_raw_only['subject_id']}")
print(f"  Trial: {sample_raw_only['trial_id']}")
print(f"  Origin: {sample_raw_only['experiment_origin']}")

Analizando trial solo en raw:
  Subject: leandrogori2000@gmail.com-tmt-plugin_112.csv
  Trial: NEUROPRUEBAS_2
  Origin: neuropruebas


In [13]:
# Intentemos cargar los datos crudos de un sujeto específico
# Primero veamos qué sujetos tienen trials solo en raw
print("Trials solo en raw:")
for idx, row in df_only_raw.iterrows():
    subj = row['subject_id']
    trial = row['trial_id']
    origin = row['experiment_origin']
    
    # Verificar si el sujeto tiene otros trials en ambos datasets
    interp_trials = set(df_interp[df_interp['subject_id'] == subj]['trial_id'])
    raw_trials = set(df_raw[df_raw['subject_id'] == subj]['trial_id'])
    common = interp_trials & raw_trials
    
    print(f"\n{subj[:40]}... | {trial}")
    print(f"  Origin: {origin}")
    print(f"  Trials en interp: {len(interp_trials)}, en raw: {len(raw_trials)}, comunes: {len(common)}")
    print(f"  Este trial está en interp: {trial in interp_trials}")

Trials solo en raw:

leandrogori2000@gmail.com-tmt-plugin_112... | NEUROPRUEBAS_2
  Origin: neuropruebas
  Trials en interp: 19, en raw: 20, comunes: 19
  Este trial está en interp: False

32d611d1-ad54-4835-99ed-460a443df14c... | NEUROPRUEBAS_3
  Origin: neuropruebas
  Trials en interp: 7, en raw: 7, comunes: 6
  Este trial está en interp: False

48482d6a-a3ca-4c42-bdab-9462f26e6ff8... | NEUROPRUEBAS_17
  Origin: neuropruebas
  Trials en interp: 11, en raw: 12, comunes: 11
  Este trial está en interp: False

4657de88-e8fe-4eca-b171-366b1822d2b3... | NEUROPRUEBAS_20
  Origin: neuropruebas
  Trials en interp: 18, en raw: 18, comunes: 17
  Este trial está en interp: False

b17c4fc1-95ed-47b9-8dd5-72ee8705a62c... | NEUROPRUEBAS_21
  Origin: neuropruebas
  Trials en interp: 18, en raw: 19, comunes: 18
  Este trial está en interp: False

3cf3fc83-1ee9-4389-9978-fa5845035019... | NEUROPRUEBAS_21
  Origin: neuropruebas
  Trials en interp: 15, en raw: 16, comunes: 15
  Este trial está en inter

In [14]:
# Lo mismo para trials solo en interpolado
print("Trials solo en interpolado (mostrando primeros 10):")
for idx, row in df_only_interp.head(10).iterrows():
    subj = row['subject_id']
    trial = row['trial_id']
    origin = row['experiment_origin']
    
    # Verificar si el sujeto tiene otros trials en ambos datasets
    interp_trials = set(df_interp[df_interp['subject_id'] == subj]['trial_id'])
    raw_trials = set(df_raw[df_raw['subject_id'] == subj]['trial_id'])
    common = interp_trials & raw_trials
    
    print(f"\n{subj[:40]}... | {trial}")
    print(f"  Origin: {origin}")
    print(f"  Trials en interp: {len(interp_trials)}, en raw: {len(raw_trials)}, comunes: {len(common)}")
    print(f"  Este trial está en raw: {trial in raw_trials}")

Trials solo en interpolado (mostrando primeros 10):

1e200df4-7a6b-4aab-b96b-8e8c8497716a... | NEUROPRUEBAS_14
  Origin: neuropruebas
  Trials en interp: 18, en raw: 17, comunes: 17
  Este trial está en raw: False

9b8b167b-7e64-4347-a579-ad8c2ba4025f... | NEUROPRUEBAS_20
  Origin: neuropruebas
  Trials en interp: 18, en raw: 17, comunes: 17
  Este trial está en raw: False

b1368f7d-47b5-45dd-83f4-5167eca84158... | NEUROPRUEBAS_2
  Origin: neuropruebas
  Trials en interp: 11, en raw: 10, comunes: 10
  Este trial está en raw: False

1aadc486-9800-46c6-93f6-13aa11168a00... | NEUROPRUEBAS_12
  Origin: neuropruebas
  Trials en interp: 12, en raw: 11, comunes: 11
  Este trial está en raw: False

b4dc8afd-c6ae-4335-9694-217a9e6f2e05... | NEUROPRUEBAS_5
  Origin: neuropruebas
  Trials en interp: 11, en raw: 9, comunes: 9
  Este trial está en raw: False

b4dc8afd-c6ae-4335-9694-217a9e6f2e05... | NEUROPRUEBAS_12
  Origin: neuropruebas
  Trials en interp: 11, en raw: 9, comunes: 9
  Este trial e

### Hipótesis Refinada

Observación clave: El validador en el mapper es **MÁS restrictivo** con interpolación (`interpolate and len(set(times)) < 2`), entonces no puede explicar por qué interpolación AGREGA trials.

La diferencia debe provenir de etapas POSTERIORES en el pipeline:
1. **Cálculo de `speed_threshold`**: Requiere tiempos estrictamente monotónicos para calcular velocidades
2. **Cálculo de métricas**: Puede fallar con `NonMonotonicTimeError` o `InvalidSpeedError`

Con interpolación, los timestamps se regeneran a intervalos fijos (60Hz), eliminando:
- Timestamps duplicados que causan `NonMonotonicTimeError`
- Intervalos muy pequeños que causan velocidades extremas (`InvalidSpeedError`)

In [15]:
# Cargar las configuraciones de cada análisis
import json

interp_config_path = Path(f'/home/gianluca/Research/datapruebas_analysis/data/hand_analysis/{INTERPOLATED_DATE}/configuration.json')
raw_config_path = Path(f'/home/gianluca/Research/datapruebas_analysis/data/hand_analysis/{RAW_DATE}/configuration.json')

with open(interp_config_path) as f:
    interp_config = json.load(f)
    
with open(raw_config_path) as f:
    raw_config = json.load(f)

print("Configuración INTERPOLADO:")
for k, v in interp_config.items():
    if k not in ['git_info', 'train_subject_ids', 'eval_subject_ids']:
        print(f"  {k}: {v}")

print("\nConfiguración RAW:")
for k, v in raw_config.items():
    if k not in ['git_info', 'train_subject_ids', 'eval_subject_ids']:
        print(f"  {k}: {v}")

Configuración INTERPOLADO:
  timestamp: 2026-01-16_19-20-02
  correct_targets_minimum: 10
  cut_criteria: MINIMUM_TARGETS
  consecutive_points: 5
  calculate_crosses: False
  interpolate_trajectory: True
  random_state: 78
  test_size: 1
  metadata_path: /home/gianluca/Research/datapruebas_analysis/data/metadata/metadata.csv
  is_new_split: False
  old_split_config_date: None
  git_branch: refresh_rate_analysis
  git_commit: 661e5cbfe078a75c69a7f0fbbb9cb64a3428887f
  git_msg: Add validation for interpolation settings in compare_interpolation_analysis
  git_dirty: True

Configuración RAW:
  timestamp: 2026-01-16_18-57-15
  correct_targets_minimum: 10
  cut_criteria: MINIMUM_TARGETS
  consecutive_points: 5
  calculate_crosses: False
  interpolate_trajectory: False
  random_state: 78
  test_size: 1
  metadata_path: /home/gianluca/Research/datapruebas_analysis/data/metadata/metadata.csv
  is_new_split: False
  old_split_config_date: None
  git_branch: refresh_rate_analysis
  git_commit: 31

In [16]:
# Análisis clave: ¿Los sujetos con trials únicos en un dataset tienen speed_threshold en ambos?
print("Sujetos con trials SOLO en interpolado:")
print("=" * 60)
for subj in subjects_only_interp:
    has_interp_st = subj in st_interp.index
    has_raw_st = subj in st_raw.index
    
    interp_trials = len(df_interp[df_interp['subject_id'] == subj])
    raw_trials = len(df_raw[df_raw['subject_id'] == subj])
    
    status = "✓ ambos" if (has_interp_st and has_raw_st) else ("solo interp" if has_interp_st else "solo raw")
    print(f"{subj[:40]}... | interp:{interp_trials} raw:{raw_trials} | speed_threshold: {status}")

print("\n" + "=" * 60)
print("Sujetos con trials SOLO en raw:")
print("=" * 60)
for subj in subjects_only_raw:
    has_interp_st = subj in st_interp.index
    has_raw_st = subj in st_raw.index
    
    interp_trials = len(df_interp[df_interp['subject_id'] == subj])
    raw_trials = len(df_raw[df_raw['subject_id'] == subj])
    
    status = "✓ ambos" if (has_interp_st and has_raw_st) else ("solo interp" if has_interp_st else "solo raw")
    print(f"{subj[:40]}... | interp:{interp_trials} raw:{raw_trials} | speed_threshold: {status}")

Sujetos con trials SOLO en interpolado:
1e200df4-7a6b-4aab-b96b-8e8c8497716a... | interp:18 raw:17 | speed_threshold: ✓ ambos
9b8b167b-7e64-4347-a579-ad8c2ba4025f... | interp:18 raw:17 | speed_threshold: ✓ ambos
b1368f7d-47b5-45dd-83f4-5167eca84158... | interp:11 raw:10 | speed_threshold: ✓ ambos
1aadc486-9800-46c6-93f6-13aa11168a00... | interp:12 raw:11 | speed_threshold: ✓ ambos
b4dc8afd-c6ae-4335-9694-217a9e6f2e05... | interp:11 raw:9 | speed_threshold: ✓ ambos
32d611d1-ad54-4835-99ed-460a443df14c... | interp:7 raw:7 | speed_threshold: ✓ ambos
26815196-070f-4f94-a765-5bbecabea52c... | interp:13 raw:12 | speed_threshold: ✓ ambos
6838a4c2-ce67-4bd1-931f-2227776ce2a4... | interp:13 raw:12 | speed_threshold: ✓ ambos
b65df3e7-91c6-40b7-a732-49622bfa8684... | interp:17 raw:16 | speed_threshold: ✓ ambos
4657de88-e8fe-4eca-b171-366b1822d2b3... | interp:18 raw:18 | speed_threshold: ✓ ambos
43fab341-f3f2-4c51-a9e6-81cd008f96d9... | interp:16 raw:14 | speed_threshold: ✓ ambos
ee1e2ab2-d64e-4e0

In [17]:
# Cargar análisis completos (incluyendo trials inválidos) para ver diferencias
interp_full_path = Path(f'/home/gianluca/Research/datapruebas_analysis/data/hand_analysis/{INTERPOLATED_DATE}/analysis.csv')
raw_full_path = Path(f'/home/gianluca/Research/datapruebas_analysis/data/hand_analysis/{RAW_DATE}/analysis.csv')

df_interp_full = pd.read_csv(interp_full_path, on_bad_lines='warn')
df_raw_full = pd.read_csv(raw_full_path, on_bad_lines='warn')

print(f"Análisis completo INTERPOLADO: {len(df_interp_full):,} filas")
print(f"  - Válidos: {df_interp_full['is_valid'].sum():,}")
print(f"  - Inválidos: {(~df_interp_full['is_valid']).sum():,}")

print(f"\nAnálisis completo RAW: {len(df_raw_full):,} filas")
print(f"  - Válidos: {df_raw_full['is_valid'].sum():,}")
print(f"  - Inválidos: {(~df_raw_full['is_valid']).sum():,}")

# Razones de invalidez
print("\n" + "=" * 60)
print("Razones de invalidez - INTERPOLADO:")
if 'invalid_cause' in df_interp_full.columns:
    print(df_interp_full[~df_interp_full['is_valid']]['invalid_cause'].value_counts())

print("\nRazones de invalidez - RAW:")
if 'invalid_cause' in df_raw_full.columns:
    print(df_raw_full[~df_raw_full['is_valid']]['invalid_cause'].value_counts())

Análisis completo INTERPOLADO: 7,508 filas
  - Válidos: 5,668
  - Inválidos: 1,840

Análisis completo RAW: 7,508 filas
  - Válidos: 5,654
  - Inválidos: 1,854

Razones de invalidez - INTERPOLADO:
invalid_cause
CUT_CRITERIA_ERROR    1577
INVALID_LENGTH         263
Name: count, dtype: int64

Razones de invalidez - RAW:
invalid_cause
CUT_CRITERIA_ERROR    1561
INVALID_LENGTH         293
Name: count, dtype: int64


In [18]:
# Comparar validez de trials entre configs
# Crear conjuntos de (subject_id, trial_id, is_valid)
interp_validity = {(r['subject_id'], r['trial_id']): r['is_valid'] 
                   for _, r in df_interp_full.iterrows()}
raw_validity = {(r['subject_id'], r['trial_id']): r['is_valid'] 
                for _, r in df_raw_full.iterrows()}

# Trials comunes
common_keys = set(interp_validity.keys()) & set(raw_validity.keys())
print(f"Trials presentes en ambos análisis completos: {len(common_keys):,}")

# Trials donde la validez cambió
validity_changed = []
for key in common_keys:
    if interp_validity[key] != raw_validity[key]:
        validity_changed.append({
            'subject_id': key[0],
            'trial_id': key[1],
            'valid_interp': interp_validity[key],
            'valid_raw': raw_validity[key]
        })

df_changed = pd.DataFrame(validity_changed)
print(f"\nTrials donde cambió la validez: {len(df_changed)}")
if len(df_changed) > 0:
    print(f"  - Pasó de inválido (raw) a válido (interp): {(~df_changed['valid_raw'] & df_changed['valid_interp']).sum()}")
    print(f"  - Pasó de válido (raw) a inválido (interp): {(df_changed['valid_raw'] & ~df_changed['valid_interp']).sum()}")

Trials presentes en ambos análisis completos: 7,508

Trials donde cambió la validez: 34
  - Pasó de inválido (raw) a válido (interp): 24
  - Pasó de válido (raw) a inválido (interp): 10


In [19]:
# Analizar los trials que cambiaron de validez
if len(df_changed) > 0:
    # Agregar invalid_cause de ambos análisis
    df_changed['invalid_cause_interp'] = df_changed.apply(
        lambda r: df_interp_full[(df_interp_full['subject_id'] == r['subject_id']) & 
                                  (df_interp_full['trial_id'] == r['trial_id'])]['invalid_cause'].values[0]
        if not r['valid_interp'] else None, axis=1
    )
    df_changed['invalid_cause_raw'] = df_changed.apply(
        lambda r: df_raw_full[(df_raw_full['subject_id'] == r['subject_id']) & 
                               (df_raw_full['trial_id'] == r['trial_id'])]['invalid_cause'].values[0]
        if not r['valid_raw'] else None, axis=1
    )
    
    print("Trials que pasaron de INVÁLIDO (raw) a VÁLIDO (interp):")
    became_valid = df_changed[~df_changed['valid_raw'] & df_changed['valid_interp']]
    print(f"  Total: {len(became_valid)}")
    if len(became_valid) > 0:
        print(f"  Causas de invalidez en raw:")
        print(became_valid['invalid_cause_raw'].value_counts())
    
    print("\nTrials que pasaron de VÁLIDO (raw) a INVÁLIDO (interp):")
    became_invalid = df_changed[df_changed['valid_raw'] & ~df_changed['valid_interp']]
    print(f"  Total: {len(became_invalid)}")
    if len(became_invalid) > 0:
        print(f"  Causas de invalidez en interp:")
        print(became_invalid['invalid_cause_interp'].value_counts())

Trials que pasaron de INVÁLIDO (raw) a VÁLIDO (interp):
  Total: 24
  Causas de invalidez en raw:
invalid_cause_raw
CUT_CRITERIA_ERROR    24
Name: count, dtype: int64

Trials que pasaron de VÁLIDO (raw) a INVÁLIDO (interp):
  Total: 10
  Causas de invalidez en interp:
invalid_cause_interp
CUT_CRITERIA_ERROR    10
Name: count, dtype: int64


## Conclusiones

### Hallazgos Clave

**34 trials cambiaron su estado de validez** entre raw e interpolado:
- **24 trials** pasaron de inválido (raw) a válido (interpolado)
- **10 trials** pasaron de válido (raw) a inválido (interpolado)
- **Efecto neto**: +14 trials

### Causa Identificada: `CUT_CRITERIA_ERROR`

Sorprendentemente, **TODOS los cambios** se deben a `CUT_CRITERIA_ERROR`, NO a `INVALID_LENGTH`.

Esto significa que:
- La validación del mapper (`len(set(times)) < 2`) NO es el factor diferenciador
- La diferencia está en cómo se detectan los "targets correctamente tocados"

### Mecanismo

El criterio `MINIMUM_TARGETS` usa `CORRECT_THRESHOLD=10` píxeles para determinar si un target fue tocado correctamente. La interpolación afecta esto porque:

1. **Remuestreo a 60Hz**: Genera puntos a intervalos regulares
2. **Cambio en la trayectoria**: Puntos intermedios pueden estar más cerca o más lejos de los targets
3. **Detección de targets**: Un trial puede pasar de "suficientes targets tocados" a "insuficientes" (o viceversa)

### Speed Threshold

- Todos los sujetos tienen `speed_threshold` en ambos datasets
- La interpolación reduce el `speed_threshold` en promedio ~24% (mediana: -22.8%)
- Esto es esperado: a 60Hz los intervalos de tiempo son más regulares, resultando en velocidades más bajas

### Implicaciones

1. **La interpolación NO solo afecta timestamps**: También modifica la geometría de la trayectoria
2. **Efecto neto positivo (+14 trials)**: Más trials pasan el criterio de targets con interpolación
3. **El efecto es a nivel de trial, no de sujeto**: Todos los sujetos afectados tienen trials en ambos datasets